Sources/SwiftProtobuf/BinaryReverseEncoder.swift

```diff
@@ -0,0 +1,156 @@
+        pointer = pointer.advanced(by: -bytes)
+    }
+
+    @discardableResult
+    private mutating func prepend(contentsOf bufferPointer: UnsafeRawBufferPointer) -> Int {
+        let count = bufferPointer.count
+	consume(count)
+        if let baseAddress = bufferPointer.baseAddress, count > 0 {
+            pointer.copyMemory(from: baseAddress, byteCount: count)
+        }
+        return count
+    }
+
+    mutating func appendUnknown(data: Data) {
+        prepend(contentsOf: data)
+    }
+
+    mutating func startField(fieldNumber: Int, wireFormat: WireFormat) {
+        startField(tag: FieldTag(fieldNumber: fieldNumber, wireFormat: wireFormat))
+    }
+
+    mutating func startField(tag: FieldTag) {
+        putVarInt(value: UInt64(tag.rawValue))
+    }
+
+    mutating func putVarInt(value: UInt64) {
+        if value > 127 {
+	    putVarInt(value: value >> 7)
+            prepend(UInt8(value & 0x7f | 0x80))
+        } else {
+            prepend(UInt8(value))
+	}
+    }
+
+    mutating func putVarInt(value: Int64) {
+        putVarInt(value: UInt64(bitPattern: value))
+    }
+
+    mutating func putVarInt(value: Int) {
+        putVarInt(value: Int64(value))
+    }
+
+    mutating func putZigZagVarInt(value: Int64) {
+        let coded = ZigZag.encoded(value)
+        putVarInt(value: coded)
+    }
+
+    mutating func putBoolValue(value: Bool) {
+        prepend(value ? 1 : 0)
+    }
+
+    mutating func putFixedUInt64(value: UInt64) {
+        var v = value.littleEndian
+        let n = MemoryLayout<UInt64>.size
+	consume(n)
+        pointer.copyMemory(from: &v, byteCount: n)
+    }
+
+    mutating func putFixedUInt32(value: UInt32) {
+        var v = value.littleEndian
+        let n = MemoryLayout<UInt32>.size
+	consume(n)
+        pointer.copyMemory(from: &v, byteCount: n)
+    }
+
+    mutating func putFloatValue(value: Float) {
+        let n = MemoryLayout<Float>.size
+        var v = value.bitPattern.littleEndian
+	consume(n)
+        pointer.copyMemory(from: &v, byteCount: n)
+    }
+
+    mutating func putDoubleValue(value: Double) {
+        let n = MemoryLayout<Double>.size
+        var v = value.bitPattern.littleEndian
+	consume(n)
+        pointer.copyMemory(from: &v, byteCount: n)
+    }
+
+    // Write a string field, including the leading index/tag value.
+    mutating func putStringValue(value: String) {
+        let utf8 = value.utf8
+        // If the String does not support an internal representation in a form
+        // of contiguous storage, body is not called and nil is returned.
+        let isAvailable = utf8.withContiguousStorageIfAvailable { (body: UnsafeBufferPointer<UInt8>) -> Int in
+            let r = prepend(contentsOf: UnsafeRawBufferPointer(body))
+            putVarInt(value: body.count)
+	    return r
+        }
+        if isAvailable == nil {
+            precondition(false)
+            let count = utf8.count
+            putVarInt(value: count)
```

Comment:
This seems backwards, but given the precondition, I'm guessing you aren't expecting this path to be needed?

Reply:
Fixed.

Reply:
(I tested this locally by commenting out the contiguous handling and re-running the tests.   Everything still seems fine.)